include/imageinfo.hpp

```diff
@@ -1118,82 +1115,95 @@ inline bool try_tga(ReadInterface &ri, size_t length, ImageInfo &info) {
-		{ kFormatJpx,   {try_jp2_jpx, Format::kFormatJp2}},
-		{ kFormatKtx,       {try_ktx, Format::kFormatUnknown}},
-		{ kFormatPng,       {try_png, Format::kFormatUnknown}},
-		{ kFormatPsd,       {try_psd, Format::kFormatUnknown}},
-		{ kFormatQoi,       {try_qoi, Format::kFormatUnknown}},
-		{ kFormatTga,       {try_tga, Format::kFormatUnknown}},
-		{kFormatTiff,      {try_tiff, Format::kFormatUnknown}},
-		{kFormatWebp,      {try_webp, Format::kFormatUnknown}},
-	};
-	constexpr size_t
-		dl_len = std::size(dl),
-		byteFlagsNeeded = dl_len / 8 + ((dl_len % 8) > 0 ? 1 : 0);
-
-	static_assert(dl_len == Format::FORMAT_COUNT, "Inconsistent list of imageinfo detectors, dl != Format");
-
     size_t length = ri.length();
 
-	uint8_t tried[byteFlagsNeeded] = { 0 };
+    constexpr DetectorInfo dl[] = {
+        {kFormatAvif, kDetectorIndexAvifHeic, try_avif_heic},
+        { kFormatBmp,      kDetectorIndexBmp,       try_bmp},
+        { kFormatCur,   kDetectorIndexCurIco,   try_cur_ico},
+        { kFormatDds,      kDetectorIndexDds,       try_dds},
+        { kFormatGif,      kDetectorIndexGif,       try_gif},
+        { kFormatHdr,      kDetectorIndexHdr,       try_hdr},
+        {kFormatHeic, kDetectorIndexAvifHeic, try_avif_heic},
+        {kFormatIcns,     kDetectorIndexIcns,      try_icns},
+        { kFormatIco,   kDetectorIndexCurIco,   try_cur_ico},
+        { kFormatJp2,   kDetectorIndexJp2Jpx,   try_jp2_jpx},
+        {kFormatJpeg,      kDetectorIndexJpg,       try_jpg},
+        { kFormatJpx,   kDetectorIndexJp2Jpx,   try_jp2_jpx},
+        { kFormatKtx,      kDetectorIndexKtx,       try_ktx},
+        { kFormatPng,      kDetectorIndexPng,       try_png},
+        { kFormatPsd,      kDetectorIndexPsd,       try_psd},
+        { kFormatQoi,      kDetectorIndexQoi,       try_qoi},
+        {kFormatTiff,     kDetectorIndexTiff,      try_tiff},
+        {kFormatWebp,     kDetectorIndexWebp,      try_webp},
+        { kFormatTga,      kDetectorIndexTga,       try_tga},
+    };
+    static_assert(FORMAT_COUNT == countof(dl), "FORMAT_COUNT != countof(dl)");
+
+    bool tried[DETECTOR_COUNT] = {false};
 
     ImageInfo info;
 
     if (most_likely_format != Format::kFormatUnknown) {
-		most_likely_format = Format(most_likely_format - 1);
-		const auto &detectorAndOrigin = std::get<1>(dl[most_likely_format]);
-		{
-			Format origin = std::get<1>(detectorAndOrigin);
-			if(origin != Format::kFormatUnknown)
-				most_likely_format = Format(origin - 1);
-		}
-		auto byteIdx = most_likely_format / 8;
-		tried[byteIdx] |= (1 << (most_likely_format - (8 * byteIdx)));
-        const auto &detector = std::get<0>(detectorAndOrigin);
-        if (detector(ri, length, info)) {
+        auto detector = dl[most_likely_format - 1];
+        tried[detector.index] = true;
+        if (detector.detect(ri, length, info) &&
+            (!must_be_one_of_likely_formats ||
+             (must_be_one_of_likely_formats && info.format() == most_likely_format))) {
```

Comment:
The `||` can already do a check for `must_be_one_of_likely_formats`, the if statement can be simplified to
```c++
if (detector.detect(ri, length, info) &&
            (!must_be_one_of_likely_formats ||
             info.format() == most_likely_format))
```
If `must_be_one_of_likely_formats` is false, it short circuits and the final result is true.
If `must_be_one_of_likely_formats` is true, the first operand will be false, and it starts to look into the 2nd operand, which already indicates that `must_be_one_of_likely_formats` is true.